Custom delimiter support

collingo opened the pull request:
Hi Nicolas,

Firstly thanks for open sourcing this. I came across it in my twitter feed this morning, just as I was preparing to write a customisable input for tags myself! The timing was impeccable :)

My requirements were a little different though, I needed spaces in my tags and for completion/output to work with commas. I have modified things slightly to allow custom delimiters to be passed on initialisation in the form...

```
insignia(input, {
  delimiter: {
    char: ',',
    code: 188
  }
});
```

Here's a pull request if it's of any use...
- added `delimiter` to options to allow overriding of space during completion/output
  - I opted to pass the keycode to save the need for a key->code map in the source
- cleaned `.tags()` and `.value()` output as it always included an empty string
- added `jshint` to `dev-dependencies` which was missing from `package.json` and required for the build step

If you have any thoughts or would prefer some changes let me know. Or just close as you see fit :)

Cheers, Nick

bevacqua commented:
I think I would just use `String.fromCharCode` to save the consumer from passing the character code in themselves [_(but see this SO answer)_](http://stackoverflow.com/a/5829387/389745).

Besides that, what are your thoughts on multi-character delimiters? The simplest case being something like 2 spaces.

Don't worry about the other fixes, I'll take care of those, you can remove them.

bevacqua commented:
Fixed the issues you've reported @ `1.3.1`

danreeves commented:
This is probably exactly as I would have implemented #2 . I agree with @bevacqua that not passing in character codes would be much better.

Any thoughts on the delimiter supporting an array?

Usecase: I want it to become a tag on space or comma:

```
delimiter:  [' ', ',']
```

bevacqua commented:
I want to stay away from something like that because it's just asking for trouble. Makes it harder to parse for no gain in my opinion. Having a single customizable delimiter makes sense because you may want spaces in your tags, for example.

Probably even the multi character delimiter doesn't make much sense, come to think of it.

Let's go for the single character delimiter. @collingo can you get rid of the `code` thing?

collingo commented:
I never knew about `String.fromCharCode` which seems a much more elegant solution - good call!

I think the `addEventListener` approach in commit https://github.com/bevacqua/insignia/commit/3f74e844e4555e230c3c3ee77f3e7b25711713fc is a nice flexible approach for having multiple tag completion keys.

To that end I have updated the PR with the following...
- simplified delimiter override to just a single character string
- added `keypress` handler in order to access unicodes
- throw error when delimiter is longer than single char
- updated docs with delimiter override
- removed previous hack to remove empty strings from `.tags()` and `.value()` output as requested

I'm pretty sure the `keydown` and `keypress` methods should cover all required cases but might be worth a double check.

bevacqua commented:
Cool. Made a bunch of comments. Could you review those and I'll give it a try later? Thanks!

collingo commented:
Sure, no worries

bevacqua commented:
Also when you're done, if you could please do:

``` shell
git reset HEAD~N # where N == number of commits you've made so far
git add .
git commit -am "Custom delimiter support"
git push --force
```

That'll squash the commits into a single one. ^^

collingo commented:
No problem, I will squash once you give the thumbs up.

Just one comment from me on the `e.preventDefault()` issue.

bevacqua commented:
Gave it a try, it's fine, but since you removed the `} else { return; }` in the `keydown` handler it wasn't working for you _(presumably)_. Add this at the end of the `keydown` event handler:

``` js
else {
  return;
}

e.preventDefault();
return false;
```

Also, you haven't removed the following piece:

``` js
if (!values[values.length-1].length) {
  values.pop();
}
```

Feel free to squash after these changes :+1:

collingo committed:
Custom delimiter support

collingo commented:
Yup, you were spot on. Squash all done :)

Thanks for your time on this, it's a great little lib

bevacqua commented:
Released as `1.4.0`. Thanks for working on this!
